sympy/matrices/matrices.py

```diff
@@ -2479,10 +2479,11 @@ def do(x):
             if len(args) == 3 and isinstance(args[2], Callable):
                 op = args[2]
                 flat_list = []
-                for i in range(rows):
+                rows_range = [cls._sympify(i) for i in range(rows)]
+                cols_range = [cls._sympify(j) for j in range(cols)]
```

Comment:
or `ix = [cls._sympify(i) for i in range(max(row, cols))]` and then index `ix[i]` or `ix[j]` as needed.